aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu

```diff
@@ -0,0 +1,160 @@
+#include <ATen/ATen.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/InitialTensorOptions.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/SparseCsrTensorImpl.h>
+#include <ATen/SparseCsrTensorUtils.h>
+#include <ATen/SparseTensorUtils.h>
+#include <ATen/WrapDimUtilsMulti.h>
+#include <ATen/native/BinaryOps.h>
+#include <ATen/native/Resize.h>
+#include <algorithm>
+
+#include <cuda_runtime.h>
+#include <type_traits>
+
+#include <THC/THCTensorMathPointwise.cuh>
+#include <THC/THCThrustAllocator.cuh>
+
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAUtils.h>
+#include <c10/cuda/CUDACachingAllocator.h>
+#include <ATen/native/sparse/cuda/SparseCUDABlas.cuh>
+
+#include <thrust/device_ptr.h>
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/sequence.h>
+
+namespace at {
+namespace native {
+
+using namespace at::sparse_csr;
+// certain utiliy functions are usable from sparse COO.
+using namespace at::sparse;
+
+Tensor& add_out_dense_sparse_csr_cuda(
+    Tensor& output,
+    const Tensor& dense,
+    const SparseCsrTensor& src,
+    const Scalar& alpha) {
+  TORCH_INTERNAL_ASSERT(dense.layout() == kStrided);
+  TORCH_INTERNAL_ASSERT(src.is_sparse_csr());
+  TORCH_INTERNAL_ASSERT(dense.is_cuda());
+
+  TORCH_CHECK(
+      output.is_contiguous(),
+      "out argument must be contiguous, but got: ",
+      output.suggest_memory_format());
+  TORCH_CHECK(
+      output.is_cuda(),
+      "add: expected 'out' to be CUDA tensor, but got tensor on device: ",
+      output.device());
+
+  TORCH_CHECK(
+      src.is_cuda(),
+      "add: expected 'other' to be a CUDA tensor, but got tensor on device: ",
+      src.device());
+
+  TORCH_CHECK(
+      dense.sizes().equals(src.sizes()),
+      "add: expected 'self' and 'other' to have same size, but self has size ",
+      dense.sizes(),
+      " while other has size ",
+      src.sizes(),
+      " (FYI: op2-sparse addition does not currently support broadcasting)");
+
+  auto commonDtype = promoteTypes(dense.scalar_type(), src.scalar_type());
+  TORCH_CHECK(
+      canCast(commonDtype, output.scalar_type()),
+      "Can't convert result type ",
+      commonDtype,
+      " to output ",
+      output.scalar_type(),
+      " in add operation");
+
+  Tensor src_values = src.values();
+  Tensor src_crow_indices = src.crow_indices();
+  Tensor src_col_indices = src.col_indices();
+
+  resize_output(output, dense.sizes());
+
+  Tensor resultBuffer = output;
+  Tensor valuesBuffer = src_values.to(commonDtype);
+  if (output.scalar_type() != commonDtype) {
+    resultBuffer = dense.to(commonDtype);
+  } else if (!is_same_tensor(output, dense)) {
+    resultBuffer.copy_(dense);
+  }
+  AT_DISPATCH_ALL_TYPES(
+      commonDtype,
+      "add_out_op2_sparse_csr",
+      [&valuesBuffer, &resultBuffer, &alpha, &src_crow_indices, &src_col_indices]() {
+        AT_DISPATCH_INDEX_TYPES(
```

Comment:
It's because operations like matvec and addmm, the CPU implementation relies on MKL, currently the APIs used here only works with int32 indices and when MKL is not enabled custom kernels for any 32 or 64 bits indices are used. This is better detailed here https://pearu.github.io/csr_tensor_support.html